app/apollo/resolvers/group.js

```diff
@@ -485,42 +537,50 @@ const groupResolvers = {
       const user = whoIs(me);
 
       try {
-        logger.info({ req_id, user, org_id, groupUuids, clusterId }, `${queryName} validating`);
-
-        await validAuth(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context);
+        logger.info({req_id, user, org_id, clusterId, groupUuids}, `${queryName} validating`);
 
         validateString( 'org_id', org_id );
         validateString( 'clusterId', clusterId );
         groupUuids.forEach( value => validateString( 'groupUuids', value ) );
 
-        const groups = await models.Group.find({ org_id, uuid: { $in: groupUuids } });
+        const groups = await getAllowedResources(me, org_id, ACTIONS.MANAGE, TYPES.GROUP, queryName, context, null, groupUuids);
+        logger.info({req_id, user, org_id, clusterId, groupUuids}, `${queryName} retrieved allowed resources`);
+
         if (groups.length != groupUuids.length) {
           throw new NotFoundError(context.req.t('One or more of the passed group uuids were not found'));
         }
 
+        if (groups.length < 1) { throw new NotFoundError(context.req.t('None of the passed group uuids were found')); }
+
         groupUuids = _.map(groups, 'uuid');
 
+        let cluster = await models.Cluster.findOne({
+          org_id,
+          cluster_id: clusterId
+        }).lean({ virtuals: true });
+        logger.info({req_id, user, org_id, clusterId, groupUuids}, `${queryName} validating - found: ${!!cluster}`);
+
+        const identifiers = cluster ? [clusterId, cluster.registration.name || cluster.name] : [clusterId];
+        await validAuth(me, org_id, ACTIONS.READ, TYPES.CLUSTER, queryName, context, identifiers);
+        logger.info({req_id, user, org_id, clusterId, groupUuids}, `${queryName} validating - cluster authorized`);
+
         // Create output for graphQL plugins
         const groupObjsToAdd = _.map(groups, (group)=>{
           return {
             uuid: group.uuid,
             name: group.name,
           };
         });
-        const clusters = await commonClusterSearch(models, {org_id}, { limit: 0, skip: 0, startingAfter: null });
-        const clusterObjs = _.map(clusters, (cluster)=>{
-          return {
-            name: cluster.registration.name,
-            uuid: cluster.cluster_id,
-            registration: cluster.registration
-          };
-        });
         const sets = {
           groups: groupObjsToAdd,
           updated: Date.now(),
         };
-
-        logger.info({ req_id, user, org_id, groupUuids, clusterId }, `${queryName} saving`);
+        cluster = {
+          name: cluster.registration.name,
+          uuid: cluster.cluster_id,
+          registration: cluster.registration
+        };
+        logger.info({req_id, user, org_id, groupUuids, clusterId}, `${queryName} saving`);
```

Comment:
- Like earlier code, a separate `const clusterObjs` should be defined for use with pluginContext, not overwriting the `cluster` variable which is the full record from the DB.  Avoiding changing the meaning of variables aids code readability, as does maintaining consistency with other code using pluginContext.
- In this case, the code flow has been interrupted by putting the definition of the cluster for pluginContext separated from the `// Create output for graphQL plugins` code comment.  I.e. it's currently:
```
[pluginContext code comment]
[group obj definition for pluginContext]
['sets' definition for database update]
[cluster obj definition for pluginContext]
[database update]
```
I.e. they're all jumbled up for some reason.  Needs to be restored to the original order.

Reply:
Dejumbled and defined as clusterObjs

Reply:
Resolved